### PR TITLE
PHI-S Paper

### PR DESCRIPTION
**Pull request by mranzinger**

### CODE REVIEW

File: radio/radio_model.py [hunk 1]
```diff
@@ -199,6 +210,9 @@ def forward_intermediates(
             intermediates_only=intermediates_only,
             aggregation=aggregation,
         )
+
+        outputs = [(summary, self.feature_normalizer(features)) for summary, features in outputs]
+
```
Comment (gheinrich) [inline radio/radio_model.py:215]: I modified this piece of code on my end:

```
        intermediates = self.model.forward_intermediates(
            x,
            indices=indices,
            return_prefix_tokens=return_prefix_tokens,
            norm=norm,
            stop_early=stop_early,
            output_fmt=output_fmt,
            intermediates_only=intermediates_only,
            aggregation=aggregation,
        )
 
        if not intermediates_only:
            final, intermediates = intermediates

        if return_prefix_tokens:
            intermediates = [(summary, self.feature_normalizer(features)) for summary, features in intermediates]
            radio_outputs = [RadioOutput(summary, features) for (summary, features) in intermediates]
        else:
            intermediates = [self.feature_normalizer(features) for features in intermediates]
            radio_outputs = [RadioOutput(None, features) for features in intermediates]

        if intermediates_only:
            return radio_outputs
        else:
            return self.feature_normalizer(final), radio_outputs
```

File: hubconf.py [hunk 2]
```diff
@@ -120,6 +122,12 @@ def radio_model(
         adaptor.head_idx = tidx
         adaptors[adaptor_name] = adaptor
 
+    feat_norm_sd = get_prefix_state_dict(state_dict, '_feature_normalizer.')
+    feature_normalizer = None
+    if feat_norm_sd:
+        feature_normalizer = FeatureNormalizer(feat_norm_sd['mean'].shape[0], dtype=dtype)
+        feature_normalizer.load_state_dict(feat_norm_sd)
```
Comment (gheinrich) [inline hubconf.py:129]: Do we need to load the state dict here? Can this not be done at the model's  top level?

File: radio/radio_model.py (hunk 1 above)
Comment (mranzinger) [reply radio/radio_model.py:215]: What about this?
```
        if return_prefix_tokens:
            radio_outputs = [
                RadioOutput(summary, self.feature_normalizer(features)) 
                for summary, features in intermediates
            ]
            radio_outputs = [RadioOutput(summary, features) for (summary, features) in intermediates]
        else:
            radio_outputs = [
                self.feature_normalizer(features) 
                for features in intermediates
            ]

        if intermediates_only:
            return radio_outputs
        else:
            return self.feature_normalizer(final), radio_outputs
```

What is `final` in this case? Is it itself a tuple? Or is it a tensor with all of the cls tokens, registers, and patches concatenated? If the latter, then I suppose we need to extract the cls tokens and patches, and normalize the patches.

File: hubconf.py (hunk 2 above)
Comment (mranzinger) [reply hubconf.py:129]: The normalizer is somewhat in purgatory. It's not part of the `base_model` since it's something I add after training. So this `.load_state_dict` is similar to the conditioner, which is extra state that we need to load, but technically not part of the base model, but rather the `radio` model which wraps it.

File: radio/radio_model.py (hunk 1 above)
Comment (gheinrich) [reply radio/radio_model.py:227]: Maybe now the docstring should specify that a list of `RadioOutput` is return if `return_prefix_tokens==True` otherwise it will be a list of feature tensors.

**Review by gheinrich — APPROVED** [review]
